frame/staking/src/offchain_election.rs

```diff
@@ -73,42 +86,90 @@ pub(crate) const DEFAULT_LONGEVITY: u64 = 25;
 /// Returns `Ok(())` if offchain worker should happen, `Err(reason)` otherwise.
 pub(crate) fn set_check_offchain_execution_status<T: Config>(
 	now: T::BlockNumber,
-) -> Result<(), &'static str> {
+	threshold: T::BlockNumber,
+) -> Result<(), OffchainElectionError> {
 	let storage = StorageValueRef::persistent(&OFFCHAIN_HEAD_DB);
-	let threshold = T::BlockNumber::from(OFFCHAIN_REPEAT);
-
-	let mutate_stat =
-		storage.mutate::<_, &'static str, _>(|maybe_head: Option<Option<T::BlockNumber>>| {
-			match maybe_head {
-				Some(Some(head)) if now < head => Err("fork."),
-				Some(Some(head)) if now >= head && now <= head + threshold => {
-					Err("recently executed.")
-				}
-				Some(Some(head)) if now > head + threshold => {
-					// we can run again now. Write the new head.
-					Ok(now)
-				}
-				_ => {
-					// value doesn't exists. Probably this node just booted up. Write, and run
-					Ok(now)
-				}
+
+	let mutate_stat = storage.mutate(|maybe_head: Option<Option<T::BlockNumber>>| {
+		match maybe_head {
+			Some(Some(head)) if now < head => Err(OffchainElectionError::Fork),
+			Some(Some(head)) if now >= head && now <= head + threshold => {
+				Err(OffchainElectionError::TooRecent)
 			}
-		});
+			Some(Some(head)) if now > head + threshold => {
+				// we can allow again now. Write the new head.
+				Ok(now)
+			}
+			_ => {
+				// value doesn't exists. Probably this node just booted up. Write, and allow.
+				Ok(now)
+			}
+		}
+	});
 
 	match mutate_stat {
 		// all good
 		Ok(Ok(_)) => Ok(()),
 		// failed to write.
-		Ok(Err(_)) => Err("failed to write to offchain db."),
+		Ok(Err(_)) => Err(OffchainElectionError::DBWriteFailed),
 		// fork etc.
 		Err(why) => Err(why),
 	}
 }
 
+/// Storage path for the solution `call`.
+pub(crate) const OFFCHAIN_QUEUED_CALL: &[u8] = b"parity/staking-election/call";
+
+/// Save a given call OCW storage.
+pub(crate) fn save_solution<T: Config>(call: Call<T>) -> Result<(), OffchainElectionError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_QUEUED_CALL);
+	let set_outcome = storage.mutate::<_, OffchainElectionError, _>(|maybe_call| {
+		match maybe_call {
+			// under any case, write the new value. This means that we don't need to clear the
+			// solution. Each era, it will be overwritten.
+			_ => Ok(call),
```

Comment:
Why not use `storage.set(call)` instead then? The whole process should be already guarded by the `OFFCHAIN_HEAD_DB` mutex storing block numbers.

Reply:
If you want to keep `mutate` just simplify it maybe to `storage.mutate::<...>(|_| Ok(call))`

Reply:
That's what I initially did but reverted to avoid: 

The `OFFCHAIN_HEAD_DB` ensures that no two OCWs can run within 5 blocks. Let's say OCW1 starts mining the block now and 6 blocks from now, as OCW1 is done with the solution and is writing it, OCW2 gets past the   `OFFCHAIN_HEAD_DB` and attempts to read this value. I don't care if the value is unavailable, but it good to ensure with a `mutate` that OCW2 will read either the whole, correct call, or nothing. 

Probably an overengineering.

Reply:
Sounds reasonable, I was suspecting that you might have done it intentionally, cause computing the solution might take a few blocks, just wanted to confirm that.